File: fight.c
#include <curses.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h>
#include "sprites.c"

// Declaração das funções do arquivo
void clear_screen();
void fight(WINDOW*);
void player_punch();
void opponent_punch();

// Variaveis globais
int option;
int playerH = 5;
int opponentH = 5;

// Limpa a tela, partir de uma coordenada
void clear_screen(){
    move(2,0); // Move o cursor para Y=2 e X=0
    clrtobot(); // Limpa a tela a partir dessa coordenada do cursor
}

// Função principal da luta
void fight(WINDOW* win) {
    srand(time(NULL)); // Define a seed do random
    playerH = 5; // Garante os valores de vida do jogador
    opponentH = 5; // Garante os valores de vida do oponente
    int opponentAttackTime = rand() % 10; // Define o tempo de ataque do oponente
    int sprite = 1; // Define o primeiro sprite da luta
        
    for(;;){
        clear(); // Limpa a tela

        // Chama as telas de vitória ou derrota
        if(playerH == 0){
            clear();
            defeat(); 
            refresh(); 
            usleep(1000000); // Aguarda 1 segundo
            clear();
            break;
        }
        else if(opponentH == 0){
            clear();
            victory(); 
            refresh();
            usleep(1000000);
            clear();
            break;
        }

        // Desenha a barra de vida
        health_bar(playerH, opponentH);

        // Ataque do oponente
        if(opponentAttackTime == 0){ 
            clear_screen(); // Limpa a tela abaixo da barra de vida
            opponent_punch(win); // Chama o sprite de soco do oponente
            opponentAttackTime = rand() % 7; // Redefine o tempo de ataque do oponente
        }
        else 
            opponentAttackTime--; // Decrementa um do tempo de ataque do oponente
        
        nodelay(win, true); // Garante a continuidade de execução
         
        int kplayer_punch = wgetch(win); // Ataque do jogador
        if(kplayer_punch == '2' && playerH > 0){ // Verifica se o jogador ainda tem vida
            flushinp(); // Limpa o cache
            clear_screen();
            player_punch(); // Soco do jogador
        }

        // Sair da luta a qualquer momento pressionando "Q" ou "q"
        if(kplayer_punch == 'q'|| kplayer_punch == 'Q'){
            break;
        }
        kplayer_punch = -1; // Redefine o valor de ataque do jogador
        
        flushinp();
        clear_screen();
        if(sprite == 1){ // Define qual dos sprites de animação irá ser exibido
            fight_sprite_1();
            refresh();
            usleep(300000);
            sprite = 0;
        }
        else{
            fight_sprite_2();
            refresh();  
            usleep(300000);
            sprite = 1;
        }
        usleep(100000); // Aguarda um segundo
    } 
}

void player_punch(){
    int opponentDefense = rand() % 2; // Define se o oponente irá defender ou não
    
    clear();
    health_bar(playerH, opponentH);
    refresh();

    player_punch_1(); // Chama o primeiro sprite do ataque do jogador
    refresh();
    usleep(300000);

    clear_screen();
    player_punch_2(); // Chama o segundo sprite do ataque do jogador
    refresh();
    usleep(300000);

    clear_screen();
    if(opponentDefense == 0){ // Defesa do oponente
        player_punch_4();
        refresh();
        usleep(800000);

    }else{
        player_punch_3();
        refresh();
        opponentH--; // Diminui a vida do oponente
        health_bar(playerH, opponentH);
        usleep(800000);
    }
}

void opponent_punch(WINDOW* win){ 
    int pressed = -1; // Defesa do ataque do oponente

    clear();
    health_bar(playerH, opponentH);
    refresh();
        
    opponent_punch_1(); // Chama o primeiro sprite do ataque do oponente
    refresh();
    usleep(300000);
    
    clear_screen();
    opponent_punch_2(); // Chama o segundo sprite do ataque do oponente
    refresh();
    flushinp(); // Limpa o buffer de entrada do usuário

    // Aguarda um curto período para capturar a entrada do jogador
    for (int i = 0; i < 10; i++) { 
        pressed = wgetch(win);
        if (pressed == '1') break; // Se o jogador pressionar "1", interrompe o loop
        usleep(30000); // Aguarda 30ms antes de tentar novamente
    }

    clear_screen();
    if(pressed == '1'){ // Se o jogador pressionar "1" a tempo, ele defende
        flushinp();
        opponent_punch_4(); // Chama a defesa do jogador
        refresh();
        usleep(800000);
    }else{
        opponent_punch_3(); // Chama o soco no jogador
        refresh();
        playerH--; // Diminui a vida do jogador
        health_bar(playerH, opponentH);
        usleep(800000);
    }
    pressed = -1;
}

File: python-version/fight.py
import curses
import random
import time
from fighter import Player, Opponent
from sprites import (
    health_bar, victory, defeat,
    fight_sprite_1, fight_sprite_2,
    player_punch_1, player_punch_2, player_punch_3, player_punch_4,
    opponent_punch_1, opponent_punch_2, opponent_punch_3, opponent_punch_4
)
from sound import play_music, stop_music, play_sound

GLOBAL_TIME = 0.4  # Tempo global de animação entre frames


class Fight:
    def __init__(self, stdscr):
        """
        Inicializa a luta, criando instâncias do jogador e oponente.
        """
        self.stdscr = stdscr
        self.player = Player("Jogador")
        self.opponent = Opponent("Oponente")

    def animate_fight(self):
        """
        Anima a tela de luta alternando entre dois sprites.
        """
        for _ in range(2):
            self.stdscr.clear()
            fight_sprite_1(self.stdscr)
            health_bar(self.stdscr, self.player.health, self.opponent.health)
            self.stdscr.refresh()
            time.sleep(GLOBAL_TIME)
            self.stdscr.clear()
            fight_sprite_2(self.stdscr)
            health_bar(self.stdscr, self.player.health, self.opponent.health)
            self.stdscr.refresh()
            time.sleep(GLOBAL_TIME)

    def run(self):
        """
        Executa o loop principal da luta, controlando ataques, defesas e input do jogador.
        """
        stop_music()
        play_music("fightSong.mp3")  # Toca música da luta
        self.stdscr.nodelay(True)  # Modo não bloqueante para input
        opponent_attack_timer = int(random.uniform(2, 8))  # Timer para ataque do oponente
        sprite = 0
        while self.player.is_alive() and self.opponent.is_alive():
            # Alterna sprites de luta
            self.stdscr.clear()
            if sprite == 0:
                fight_sprite_1(self.stdscr)
                sprite = 1
            else:
                fight_sprite_2(self.stdscr)
                sprite = 0
            health_bar(self.stdscr, self.player.health, self.opponent.health)
            self.stdscr.refresh()
            time.sleep(GLOBAL_TIME)

            # Ataque do oponente se timer chegar a zero
            opponent_attack_timer -= 1
            if opponent_attack_timer <= 0:
                # Oponente ataca o jogador
                self.opponent.attack(self.player, stdscr=self.stdscr)
                opponent_attack_timer = int(random.uniform(2, 8))

            # Input do jogador
            key = self.stdscr.getch()
            if key == ord('2'):
                # 30% de chance do oponente defender
                opponent_defended = (random.random() < 0.3)
                self.player.attack(
                    self.opponent, stdscr=self.stdscr, defended=opponent_defended)
                curses.flushinp()  # Limpa buffer de entrada
            elif key in (ord('q'), ord('Q')):
                break
            time.sleep(0.05)
        self.stdscr.nodelay(False)
        stop_music()
        # Exibe vitória ou derrota com sprite
        if self.player.is_alive() and self.opponent.is_alive():
            # Saiu pelo 'q', não mostra vitória/derrota
            return
        self.stdscr.clear()
        if self.player.is_alive():
            play_sound("Winner.mp3")
            victory(self.stdscr)
        else:
            play_sound("Lose.mp3")
            defeat(self.stdscr)
        self.stdscr.refresh()
        time.sleep(4)
